### PR TITLE
Explicitly specify pad token id when generating tokens

### LINKED ISSUE #3536: Explicitly specify pad token id when generating tokens
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
We have added a [fallback pad token if it is not already present in the tokenizer](https://github.com/kserve/kserve/blob/11e0ab2ca4d852ec154156fad4f9d220c8768d78/python/huggingfaceserver/huggingfaceserver/model.py#L130-L131) as part of the PR #3459. But it does not explicitly specifies pad_token_id when invoking generate method which leads to huggingface using eos_token_id as the pad_token_id. The log from huggingface server is show below. To avoid this we should explicitly specify the pad_token_id
```
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
```

**What did you expect to happen:**
pad_token_id should be set accordingly.

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]
```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: huggingface-gpt2
spec:
  predictor:
    model:
      modelFormat:
        name: huggingface
      args:
        - "--model_name=gpt2"
        - "--model_id=openai-community/gpt2"
        - "--tensor_input_names=input_ids"
        - "--disable_vllm"
        - "--task=6"
      env:
        - name: CUDA_LAUNCH_BLOCKING
          value: "1"
      resources:
        limits:
          nvidia.com/gpu: "1"
        requests:
          nvidia.com/gpu: "1"
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: master
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:
We have added a [fallback pad token if it is not already present in the tokenizer](https://github.com/kserve/kserve/blob/11e0ab2ca4d852ec154156fad4f9d220c8768d78/python/huggingfaceserver/huggingfaceserver/model.py#L130-L131) as part of the PR #3459. But it does not explicitly specifies pad_token_id when invoking generate method which leads to huggingface using eos_token_id as the pad_token_id. The log from huggingface server is show below. To avoid this we should explicitly specify the pad_token_id
```
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
```
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3536

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Explicitly specify pad token id when generating tokens
```

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
There seems to be other unrelated changes

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 1]
```diff
@@ -129,6 +129,9 @@ def load(self) -> bool:
             model_id_or_path, do_lower_case=self.do_lower_case, device_map=self.device_map)
         if not self.tokenizer.pad_token:
             self.tokenizer.add_special_tokens({'pad_token': '[PAD]'})
+        if self.tokenizer.pad_token is None:
```
Comment (yuzisun) [inline python/huggingfaceserver/huggingfaceserver/model.py]: Do you mean we need to add the special token for both cases? then we do not need the conditions

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 2]
```diff
@@ -251,7 +252,7 @@ async def predict(self, input_batch: Union[BatchEncoding, InferRequest], context
             try:
                 with torch.no_grad():
                     if self.task == MLTask.text2text_generation.value or self.task == MLTask.text_generation:
-                        outputs = self.model.generate(**input_batch)
+                        outputs = self.model.generate(**input_batch, pad_token_id=self.tokenizer.pad_token_id)
```
Comment (yuzisun) [inline python/huggingfaceserver/huggingfaceserver/model.py:269]: is there a test case we can add ?

File: python/huggingfaceserver/huggingfaceserver/model.py (hunk 1 above)
Comment (sivanantha321) [reply python/huggingfaceserver/huggingfaceserver/model.py]: Sorry, my rebase got messed up. will remove the extra condition

File: python/huggingfaceserver/huggingfaceserver/model.py (hunk 2 above)
Comment (sivanantha321) [reply python/huggingfaceserver/huggingfaceserver/model.py:269]: Added

Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/model.py:269]: the test seems to be failing
https://github.com/kserve/kserve/actions/runs/8417844462/job/23047080962?pr=3535

Comment (sivanantha321) [reply python/huggingfaceserver/huggingfaceserver/model.py:269]: fixed

**Pull request closed without merging**